File: runtime-agent/policy/stage2_heuristic.go
package policy

import (
	"context"
	"regexp"
	"time"
)

// HeuristicStage2 is the zero-config inline Stage 2 classifier. It mirrors the
// control plane's app/policy/stage2_heuristic.py: a deterministic lexical /
// structural detector for prompt-injection and jailbreak attempts that runs
// inline with NO model weights, so "balanced"/"comprehensive" enforcement does
// real ML-ish detection out of the box (replacing the old noopStage2{}).
//
// When a trained ONNX model is provisioned, HTTPStage2 (an inference sidecar)
// supersedes this — see stage2_http.go. The heuristic remains the fallback so
// the agent is never stuck at Stage-1-only.
type HeuristicStage2 struct {
	signals []heuristicSignal
}

type heuristicSignal struct {
	re       *regexp.Regexp
	weight   float64
	category string
}

// Compiled once. Weights tuned so a single strong phrase lands in the high
// band (act now) and a single weak/structural signal in the uncertain band
// (escalate to Stage 3) — matching the Python heuristic.
var stage2Signals = []heuristicSignal{
	{regexp.MustCompile(`(?i)\bignore\s+(?:all\s+)?(?:previous|prior|above)\s+instructions?\b`), 0.75, "prompt_injection"},
	{regexp.MustCompile(`(?i)\bdisregard\s+(?:all\s+)?(?:previous|prior|the\s+above)\b`), 0.7, "prompt_injection"},
	{regexp.MustCompile(`(?i)\byou\s+are\s+now\b`), 0.45, "jailbreak"},
	{regexp.MustCompile(`(?i)\b(?:DAN|do\s+anything\s+now)\b`), 0.7, "jailbreak"},
	{regexp.MustCompile(`(?i)\bdeveloper\s+mode\b`), 0.55, "jailbreak"},
	{regexp.MustCompile(`(?i)\b(?:system|initial)\s+prompt\b`), 0.5, "prompt_injection"},
	{regexp.MustCompile(`(?i)\brepeat\s+(?:the\s+)?(?:text|words|everything)\s+above\b`), 0.6, "prompt_injection"},
	{regexp.MustCompile(`(?i)\bpretend\s+(?:to\s+be|you\s+are)\b`), 0.4, "jailbreak"},
	{regexp.MustCompile(`(?i)\boverride\s+(?:your\s+)?(?:safety|guidelines|rules)\b`), 0.65, "jailbreak"},
	// Structural: a long contiguous base64/hex blob often hides a payload.
	{regexp.MustCompile(`[A-Za-z0-9+/]{120,}={0,2}`), 0.35, "prompt_injection"},
}

// NewHeuristicStage2 returns the zero-config inline classifier.
func NewHeuristicStage2() *HeuristicStage2 {
	return &HeuristicStage2{signals: stage2Signals}
}

// Classify implements Stage2Engine. Returns matched with a calibrated
// confidence; the orchestrator handles routing (high → act, uncertain →
// Stage 3). Allocation-light: no work beyond the regex scans on the hot path.
func (h *HeuristicStage2) Classify(_ context.Context, in *Input, _ *CompiledPolicy) StageResult {
	start := time.Now()
	text := in.Text

	var confidence float64
	hits := 0
	byCategory := map[string]float64{}
	for _, s := range h.signals {
		if s.re.MatchString(text) {
			confidence += s.weight
			hits++
			byCategory[s.category] += s.weight
		}
	}
	if confidence > 1.0 {
		confidence = 1.0
	}
	latency := time.Since(start).Microseconds()

	if hits == 0 {
		return StageResult{
			Stage:     ExitStage2ML,
			Matched:   false,
			Action:    ActionAllowed,
			LatencyUS: latency,
		}
	}

	category := ""
	var best float64
	for c, w := range byCategory {
		if w > best {
			best, category = w, c
		}
	}
	severity := SeverityMedium
	if confidence >= 0.7 {
		severity = SeverityHigh
	}
	return StageResult{
		Stage:      ExitStage2ML,
		Matched:    true,
		Action:     ActionFlagged,
		Severity:   severity,
		Category:   category,
		RuleID:     "heuristic-stage2",
		Confidence: confidence,
		Reason:     "heuristic Stage 2 matched " + category,
		LatencyUS:  latency,
		Evidence:   map[string]any{"signals": hits, "confidence": confidence},
	}
}

File: backend/app/policy/stage2_heuristic.py
"""Stage 2 — zero-config heuristic classifier.

The ONNX engine (:mod:`app.policy.stage2_onnx`) is the high-fidelity Stage 2,
but it needs model artifacts provisioned per deployment. Until an operator
drops in a trained model, the pipeline still needs a *functional* Stage 2 so
"balanced" / "comprehensive" enforcement actually does ML-ish detection out
of the box rather than silently passing everything through (the old
``_NoopStage2`` behaviour).

This is a deterministic, dependency-free lexical/structural classifier for
prompt-injection and jailbreak attempts. It is intentionally *not* presented
as a trained model — confidences are calibrated from signal strength, and the
ONNX engine supersedes it when configured. It exists so the three-stage
pipeline is real end-to-end with no external dependencies, and so the
confidence-routing path (Stage 2 -> Stage 3 escalation) is exercised.

Signals (each contributes weighted evidence toward a 0-1 confidence):
  - Instruction-override phrases ("ignore previous instructions", …)
  - Role / persona hijack ("you are now", "DAN", "developer mode", …)
  - System-prompt exfiltration ("repeat the text above", "system prompt")
  - Encoded-payload smell (long base64/hex blobs that hide instructions)
"""

from __future__ import annotations

import re
import time

from app.policy.compiled import CompiledPolicy
from app.policy.types import PolicyInput, StageResult

# Each pattern carries (weight, category). Weights sum (capped at 1.0) into
# the confidence. Tuned so a single strong phrase lands in the "high" band
# (≥0.7) and a single weak/structural signal lands in the "uncertain" band
# (0.3-0.7) so Stage 3 escalation gets exercised.
_SIGNALS: tuple[tuple[re.Pattern[str], float, str], ...] = (
    (
        re.compile(r"\bignore\s+(?:all\s+)?(?:previous|prior|above)\s+instructions?\b", re.I),
        0.75,
        "prompt_injection",
    ),
    (
        re.compile(r"\bdisregard\s+(?:all\s+)?(?:previous|prior|the\s+above)\b", re.I),
        0.7,
        "prompt_injection",
    ),
    (re.compile(r"\byou\s+are\s+now\b", re.I), 0.45, "jailbreak"),
    (re.compile(r"\b(?:DAN|do\s+anything\s+now)\b", re.I), 0.7, "jailbreak"),
    (re.compile(r"\bdeveloper\s+mode\b", re.I), 0.55, "jailbreak"),
    (re.compile(r"\b(?:system|initial)\s+prompt\b", re.I), 0.5, "prompt_injection"),
    (
        re.compile(r"\brepeat\s+(?:the\s+)?(?:text|words|everything)\s+above\b", re.I),
        0.6,
        "prompt_injection",
    ),
    (re.compile(r"\bpretend\s+(?:to\s+be|you\s+are)\b", re.I), 0.4, "jailbreak"),
    (
        re.compile(r"\boverride\s+(?:your\s+)?(?:safety|guidelines|rules)\b", re.I),
        0.65,
        "jailbreak",
    ),
    # Structural: a long contiguous base64/hex blob often hides a payload.
    (re.compile(r"[A-Za-z0-9+/]{120,}={0,2}"), 0.35, "prompt_injection"),
)


class HeuristicStage2:
    """Stage 2 engine implementing the Stage2Engine protocol with no ML deps.

    Returns ``matched=True`` with ``action="flagged"`` and a calibrated
    confidence; the orchestrator handles routing (high -> act, uncertain ->
    Stage 3). Returns ``matched=False`` when no signal fires.
    """

    async def classify(self, *, input_: PolicyInput, policy: CompiledPolicy) -> StageResult:
        start_ns = time.perf_counter_ns()
        text = input_.text or ""

        confidence = 0.0
        hits: list[str] = []
        # Track per-category contribution so we can report the dominant one.
        by_category: dict[str, float] = {}
        for pattern, weight, category in _SIGNALS:
            if pattern.search(text):
                confidence += weight
                hits.append(pattern.pattern)
                by_category[category] = by_category.get(category, 0.0) + weight

        confidence = min(confidence, 1.0)
        latency_us = (time.perf_counter_ns() - start_ns) // 1000

        if not hits:
            return StageResult(
                stage="stage2_ml",
                mode="stage2_heuristic",
                matched=False,
                action="allowed",
                latency_us=int(latency_us),
            )

        category = max(by_category, key=lambda k: by_category[k])
        severity = "high" if confidence >= 0.7 else "medium"
        return StageResult(
            stage="stage2_ml",
            mode="stage2_heuristic",
            matched=True,
            action="flagged",
            severity=severity,
            category=category,
            rule_id="heuristic-stage2",
            confidence=round(confidence, 4),
            reason=f"heuristic Stage 2 matched {category!r} ({len(hits)} signal(s))",
            latency_us=int(latency_us),
            evidence={"signals": hits[:8], "confidence": round(confidence, 4)},
        )
